vscode/src/debugger/session.ts

```diff
@@ -144,57 +190,392 @@ export class QscDebugSession extends LoggingDebugSession {
-    const programText = await this._fileAccessor.readFileAsString(args.program);
-
-    const diagnostics = await this._compiler.checkCode(programText);
-    if (diagnostics.length > 0) {
-      log.trace("compilation failed. sending error response");
-      this.sendErrorResponse(response, {
-        id: -1,
-        format: ErrorProgramHasErrors,
-        showUser: true,
-      });
-      this.sendResponse(response);
-      return;
-    }
-
     if (args.noDebug) {
       log.trace(`Running without debugging`);
-      this.sendResponse(response);
-      await delay(LaunchDelayMS);
-      this.runWithoutDebugging(args);
+      await this.runWithoutDebugging(args);
     } else {
       log.trace(`Running with debugging`);
-      this.sendResponse(response);
-      // This is where we would start the debugger
-      await delay(LaunchDelayMS);
-      this.runWithoutDebugging(args);
+      await this.runWithDebugging(args);
     }
+    log.trace(`sending launchRequest response`);
+    this.sendResponse(response);
   }
-
-  private async runWithoutDebugging(
-    args: ILaunchRequestArguments
+  private async runWithDebugging(
+    _args: ILaunchRequestArguments
   ): Promise<void> {
-    const programText = await this._fileAccessor.readFileAsString(args.program);
-
+    const bps = this.get_breakpoint_ids();
+    this.run(bps);
+  }
+  private async run(bps: number[]): Promise<void> {
     const eventTarget = createDebugConsoleEventTarget();
-    this._compiler
-      .run(programText, args.entry ?? "", args.shots, eventTarget)
-      .then(() => {
+    const res = await this.debugService
+      .evalContinue(bps, eventTarget)
+      .catch((e) => {
+        log.info(`ending session due to error: ${e}`);
         vscode.debug.activeDebugConsole.appendLine("");
         vscode.debug.activeDebugConsole.appendLine(SimulationCompleted);
         this.sendEvent(new TerminatedEvent());
         this.sendEvent(new ExitedEvent(0));
       });
+
+    if (res) {
+      log.trace(`raising breakpoint event`);
+      const evt = new StoppedEvent(
+        "breakpoint",
+        QscDebugSession.threadID
+      ) as DebugProtocol.StoppedEvent;
+      evt.body.hitBreakpointIds = [res];
+      this.sendEvent(evt);
+
+      this.sendEvent(
+        new BreakpointEvent("changed", {
+          verified: true,
+          id: res,
+        } as DebugProtocol.Breakpoint)
+      );
+    } else {
+      this.endSession(`ending session`);
+    }
+  }
+
+  private endSession(message: string) {
+    log.trace(message);
+    vscode.debug.activeDebugConsole.appendLine("");
+    vscode.debug.activeDebugConsole.appendLine(SimulationCompleted);
+    this.sendEvent(new TerminatedEvent());
+    this.sendEvent(new ExitedEvent(0));
+  }
+
+  private async runWithoutDebugging(
+    args: ILaunchRequestArguments
+  ): Promise<void> {
+    const bps: number[] = [];
+    for (let i = 0; i < args.shots; i++) {
+      this.run(bps);
+    }
+  }
+
+  private get_breakpoint_ids(): number[] {
```

Comment:
Why is this snake_case when all the other methods are camelCase?

Reply:
I forget what language I'm in.